core: Added quick fixes for missing bundles (implements #1589)

kriegfrj opened the pull request:
This PR is an implementation of #1589.

If you have an import statement for a package/class that is not on the build path, and Bndtools can find one or more bundles that exports that package in the known set of local repositories, then it will add a "quick fixes" to the top of the available quick fixes. Similarly for fully-qualified references to unknown types. 

Bundles from the current workspace will be listed first, then from other repositories.

If you select the "quick fix" for a given bundle, then that bundle will be added to the build path in the `bnd.bnd` file for the current project.

There are some heuristics in there to try to distinguish between class names and package names where it is syntactically ambiguous (basically, in most cases it is assumed to be a package name if it starts with a lower case letter and a class name otherwise).

I tried to have a fairly comprehensive set of automated tests as well, though I know I fell short of 100% coverage.

Possible future enhancements:

# Demote suggestions to add a bundle if a different bundle is already on the build path that exports the same package.
# Add auto-completion proposals for types/imports of packages from known bundles.
# Index the exported classes in the bundles as well as the exported packages, so that we could provide type fixes/completions as well as package fixes/completions.

bjhargrave commented on bndtools.core/src/org/bndtools/core/editors/ImportPackageQuickFixProcessor.java:
```diff
@@ -0,0 +1,578 @@
+            fBndModel.saveChangesTo(fBndDoc);
+            InputStream str;
+            try {
+                str = new ByteArrayInputStream(fBndDoc.get()
```
In general, you want to avoid ByteArray{Input,Output}Stream classes. They are excessively synchronized. We have ByteBuffer{Input,Output}Stream in the aQute.lib.io package as replacements.

kriegfrj replied:
Changed to use `ByteBufferInputStream`.

bjhargrave commented on bndtools.core/src/org/bndtools/core/editors/ImportPackageQuickFixProcessor.java:
```diff
@@ -0,0 +1,578 @@
+            return fBuildPathBundles.contains(bundle);
+        }
+
+        // public boolean containsBundle(VersionedClause versionedBundle) throws CoreException {
```
Delete the commented out code.

kriegfrj replied:
Done.

bjhargrave commented on bndtools.core/src/org/bndtools/core/editors/ImportPackageQuickFixProcessor.java:
```diff
@@ -0,0 +1,578 @@
+            fContext = context;
+        }
+
+        // public void printBndFile(String msg, IFile file) {
```
delete the commented out code

kriegfrj replied:
Done.

bjhargrave commented on bndtools.core/src/org/bndtools/core/editors/ImportPackageQuickFixProcessor.java:
```diff
@@ -0,0 +1,578 @@
+        }
+    }
+
+    protected static Requirement buildRequirement(String importName) {
```
You use `protected` a lot. Unless this class is designed for subclassing which I do not see that  it is, the just use default or private access for fields and methods.

kriegfrj replied:
Done.

bjhargrave commented on bndtools.core/src/org/bndtools/core/editors/ImportPackageQuickFixProcessor.java:
```diff
@@ -0,0 +1,578 @@
+
+                return Status.OK_STATUS;
+            } catch (CoreException e) {
+                return e.getStatus();
```
I am thinking you may want to log the exception here?

kriegfrj replied:
I had assumed that the Eclipse Job framework would take care of properly reporting the error that is returned via the `IStatus` return value, but given that you asked the question I decided to verify this empirically. I have verified that this is indeed the case - not only will Eclipse log the error, but it will pop up a dialog with information about the error. If we log it manually we'll get duplicates in the error log.

bjhargrave commented on bndtools.core/src/org/bndtools/core/editors/ImportPackageQuickFixProcessor.java:
```diff
@@ -0,0 +1,578 @@
+        List<RepositoryPlugin> ps = listRepositories();
+
+        final BndBuildPathHandler handler = getBuildPathHandler(context);
+        SetMultimap<String, String> bundles = MultimapBuilder.treeKeys()
```
Don't use google collections

kriegfrj replied:
Fixed.

bjhargrave commented on bndtools.core/bnd.bnd:
```diff
@@ -10,8 +10,10 @@ Bundle-Localization: plugin
+Import-Package: \
+	junit.framework;resolution:=optional,\
 	org.apache.felix.shell;resolution:=optional,\
+	com.google.common.collect;version=16.0.0,\
```
I see that org.bndtools.templating uses google collections because the mustache templating code uses it, but it uses -conditionalpackage to include the used packages as implementation detail to avoid importing the packages causing a runtime dependencies on another bundle.

kriegfrj replied:
This is a bit moot now, as I have removed the Guava dependency by changing from the Guava multimap implementation to the Bnd one. However:

1. I would not have pulled in guava were it not for the fact that there was already another bundle in the workspace that was using it.
2. While I can understand dependency fanout being an issue for Bnd, I would argue that it is much less of an issue for Bndtools. The latter is deployed in the context of an Eclipse installation, which already has a large number of dependencies installed (and for this example in particular, Guava is already likely to be among those dependencies, so we wouldn't be adding to the clutter).

bjhargrave commented on bndtools.core/src/org/bndtools/core/editors/ImportPackageQuickFixProcessor.java:
```diff
@@ -0,0 +1,578 @@
+                            bundles.put(bsn, p.getName());
+                        }
+                    } catch (Exception e) {
+                        e.printStackTrace();
```
This is not useful in an IDE since no one may see it. Better to log.

kriegfrj replied:
Yes, that was an oversight. I have removed the try/catch and allowed the `CoreException` to propagate up to the Java core through `getCorrections()`. I have verified that Eclipse takes care of reporting the error (to its error log).

bjhargrave commented on bndtools.core/src/org/bndtools/core/editors/ImportPackageQuickFixProcessor.java:
```diff
@@ -0,0 +1,578 @@
+     * @param cap The capability object describing the package being imported.
+     * @return A user-friendly string with the BSN and version.
+     */
+    public static VersionedClause capabilityToVersion(Capability cap) {
```
This method does not appear to be used.

kriegfrj replied:
It was used in an earlier iteration, when I was toying with the idea of showing suggestions for individual available bundle versions, rather than simply suggesting the bare bsn. I left the code in in case I decided to add this functionality again in the future (similarly with `BndBuildPathHandler.containsBundle(VersionedClause)`). I removed it now though - it won't be hard to add again if needed.

kriegfrj commented on bndtools.core/src/org/bndtools/core/editors/ImportPackageQuickFixProcessor.java:
```diff
@@ -0,0 +1,578 @@
+    }
+
+    protected static class BndBuildPathHandler {
+        private final IInvocationContext fContext;
```
I copied what appears to be the Eclipse coding standard. Will change to mimic what Bndtools code is doing.

kriegfrj commented on bndtools.core/src/org/bndtools/core/editors/ImportPackageQuickFixProcessor.java:
```diff
@@ -0,0 +1,578 @@
+import org.osgi.service.repository.Repository;
+
+import com.google.common.collect.MultimapBuilder;
+import com.google.common.collect.SetMultimap;
```
I was not aware of that - I've changed it to use the Bnd implementation.

One thing to watch will be performance on workspaces with large numbers of repositories, as now the `contains()` call where this was used will now be doing a linear search. With the Guava multimap I had set it up to use a TreeSet as the "bucket" collection type for sorting and more efficient lookup. But the sorting probably isn't that important, and in most workspaces the number of repos will probably be small enough that a linear search will be fine. I didn't notice any significant performance hit in testing. However, if it becomes an issue might need to enhance Bnd's implementation so as to enable specifying alternative collection types for the buckets so that we can have one with Set/fast lookup characteristics.

kriegfrj commented on bndtools.core/bndtools.shared.bndrun:
```diff
@@ -166,6 +166,7 @@ crossplatform.bundles: \
 	${workspace.bundles},\
 	${gogo.bundles},\
 	bndtools.runtime.eclipse.applaunch,\
+	com.google.guava;version=16.0.0,\
```
This has been removed; see above comments about Guava.

kriegfrj commented on bndtools.core/bnd.bnd:
```diff
@@ -98,12 +100,16 @@ eclipse.deps: \
 	org.apache.servicemix.bundles.xmlpull;version=latest,\
-	bndtools.utils;version=project;packages=*
+	bndtools.utils;version=project;packages=*,\
+	com.google.guava;version=16.0
```
This has been removed; see above comments about Guava.

seanbright commented on bndtools.core/test/org/bndtools/core/editors/FakeFile.java:
```diff
@@ -0,0 +1,268 @@
+             * filter.indexOf('='); final String pkg = filter.substring(start + 1, filter.length() - 1) .trim(); final
+             * Collection<Capability> caps = thisRepoMap.get(pkg); retval.put(req, (caps == null ?
+             * Collections.emptyList() : caps)); i++; } return retval; } }).when(repo) .findProviders(any());
+             */
```
Delete this comment blob

kriegfrj replied:
Done.

seanbright commented on bndtools.core/src/org/bndtools/core/editors/ImportPackageQuickFixProcessor.java:
```diff
@@ -0,0 +1,520 @@
+    }
+
+    static private Requirement buildRequirement(String importName) {
+        String filter = String.format("(%s=%s)", PackageNamespace.PACKAGE_NAMESPACE, importName.trim());
```
I could be wrong, but I feel that we could use one of the `add...()` methods on `CapReqBuilder` to handle this bit. If I'm wrong feel free to ignore.

kriegfrj replied:
I have managed to change this to use `CapReqBuilder` exclusively, as you suggested.

Note that the implementation of `buildRequirement()` was originally lifted from here: https://github.com/bndtools/bndtools/blob/ca95e3ca7ba2344ec21275eea29b166d1dbc7fcc/bndtools.core/src/bndtools/views/repository/PackageSearchPanel.java#L94-L97

and here:
https://github.com/bndtools/bndtools/blob/ca95e3ca7ba2344ec21275eea29b166d1dbc7fcc/bndtools.core/src/bndtools/views/repository/PackageSearchPanel.java#L105-L110

... so maybe the search code in `bndtools.views.repository` could also be simplified in this way.

seanbright commented on bndtools.core/bnd.bnd:
```diff
@@ -1,4 +1,4 @@
-# Set javac settings from JDT prefs
+	# Set javac settings from JDT prefs
```
Revert this whitespace change

kriegfrj replied:
FYI: I don't think I explicitly made these whitespace changes - I think the Bndtools editor did while I was working on it. :smile:

I'm happy to revert, but I'm curious about the reason. Is it because you don't want unrelated changes as part of this commit? Or is it because you liked the whitespace the way it was originally?

seanbright replied:
My personal preference is that whitespace or formatting-only changes are not commingled with functional changes. It makes reviewing the functional changes easier.

seanbright commented on bndtools.core/bnd.bnd:
```diff
@@ -10,7 +10,8 @@ Bundle-Localization: plugin
 # 1. JUNit and Felix Shell are optional dependencies.
 # 2. Disable ALL Eclipse split package attributes, to ensure we import from the "aggregator" bundle(s).
-Import-Package: junit.framework;resolution:=optional,\
+Import-Package: \
```
Revert this whitespace change

kriegfrj replied:
See above.

seanbright commented on bndtools.core/bnd.bnd:
```diff
@@ -98,12 +99,15 @@ eclipse.deps: \
 	org.bndtools.headless.build.manager;version=latest,\
 	org.bndtools.versioncontrol.ignores.manager;version=latest,\
-	org.bndtools.templating; version=latest,\
+	org.bndtools.templating;version=latest,\
```
Revert this whitespace change

kriegfrj replied:
See above.

seanbright commented on bndtools.core/src/org/bndtools/core/editors/ImportPackageQuickFixProcessor.java:
```diff
@@ -0,0 +1,520 @@
+
+    // Relevance constants
+    final public static int ADD_BUNDLE = 15;
+    final public static int ADD_BUNDLE_WORKSPACE = ADD_BUNDLE + 1;
```
I would prefer the visibility keyword to be first on these.

bjhargrave replied:
accessibility keyword :-)

kriegfrj replied:
Fixed. Any chance of this getting added to the code formatter so that it automatically puts all of the modifiers in the correct order (just like it auto-magically formats so much other stuff - whether I want it to or not? 😉 )

bjhargrave commented:
Also, you will need to squash the commits to remove the intermediate states (bug fixes, addressing PR comments).

kriegfrj commented on bndtools.core/src/org/bndtools/core/editors/ImportPackageQuickFixProcessor.java:
```diff
@@ -0,0 +1,520 @@
+            }
+
+            for (Capability providerCap : values) {
+                resultSet.add(providerCap);
```
Thanks for highlighting this. I actually managed to simplify it even further by changing `searchRepositories()` to return a `Collection` rather than a `Set`, so that I can return the search results directly. This eliminates the copy completely and is this more efficient.

kriegfrj commented:
Ok, I've processed the latest round of suggestions, rebased onto the latest HEAD for master, and squashed all the commits.

I also added another feature - the suggestion will list the repo that the suggested bundle comes from. If the bundle is found in more than one repo, it will name one of the repos and then say " + n other(s)".

It would be nice if you could get Eclipse to show you additional information for quick fixes - as yet, I haven't figured out how to do this, but if it could be done easily, then it could list all of the repos explicitly. Next step after that would be to have options for particular versions.

bjhargrave commented:
@seanbright Are you ok to merge?

bjhargrave commented:
Thanks for the PR!

kriegfrj commented:
No problem!

kriegfrj commented:
Btw, I think this closes #1408 as well.
